### PR TITLE
Copy and normalize query data when cosine for thread safe

### PR DESCRIPTION
**Pull request by cydrain**
Issue: https://github.com/milvus-io/milvus/issues/26202

### CODE REVIEW

File: src/index/ivf/ivf.cc [hunk 1]
```diff
@@ -498,14 +509,32 @@ IvfIndexNode<T>::RangeSearch(const DataSet& dataset, const Config& cfg, const Bi
                     auto cur_data = (const uint8_t*)xq + index * dim / 8;
                     index_->range_search_thread_safe(1, cur_data, radius, &res, index_->nlist, bitset);
                 } else if constexpr (std::is_same<T, faiss::IndexIVFFlat>::value) {
-                    auto cur_data = (const float*)xq + index * dim;
-                    index_->range_search_without_codes_thread_safe(1, cur_data, radius, &res, index_->nlist, 0, bitset);
+                    auto cur_query = (const float*)xq + index * dim;
```
Comment (liliu-z) [inline src/index/ivf/ivf.cc:512]: Why we don't do 
```
auto cur_query = (const float*)xq + index * dim;
if (is_cosine) {
  cur_query = CopyAndNormalizeFloatVec(cur_query, dim);
}
```
And then we don't have to do any changes in the if else blocks

File: thirdparty/hnswlib/hnswlib/hnswalg.h [hunk 2]
```diff
@@ -1154,14 +1154,18 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
...
             return {};
 
+        size_t dim = *(size_t*)dist_func_param_;
+
         // do normalize for COSINE metric type
+        std::unique_ptr<float[]> query_data_norm;
         if (metric_type_ == Metric::COSINE) {
-            knowhere::NormalizeVec((float*)query_data, *((size_t*)dist_func_param_));
+            query_data_norm = knowhere::CopyAndNormalizeFloatVec((const float*)query_data, dim);
+            query_data = query_data_norm.get();
```
Comment (chasingegg) [inline thirdparty/hnswlib/hnswlib/hnswalg.h:1168]: will this operation still changes the original query_data?

File: src/index/ivf/ivf.cc [hunk 3]
```diff
@@ -423,17 +419,36 @@ IvfIndexNode<T>::Search(const DataSet& dataset, const Config& cfg, const BitsetV
...
+                    } else {
+                        index_->search_thread_safe(1, cur_query, k, distances + offset, ids + offset, nprobe,
+                                                   scann_cfg.reorder_k.value(), bitset);
+                    }
                 } else {
-                    auto cur_data = (const float*)data + index * dim;
-                    index_->search_thread_safe(1, cur_data, k, distances + offset, ids + offset, nprobe, 0, bitset);
+                    auto cur_query = (const float*)data + index * dim;
+                    if (is_cosine) {
+                        auto cur_query_norm = CopyAndNormalizeFloatVec(cur_query, dim);
```
Comment (chasingegg) [inline src/index/ivf/ivf.cc:445]: better to rename all this `cur_query_norm` to `cur_normalized_query`, it is confusing to use query_norm which could be considered as a numeric number means norm.

File: src/index/ivf/ivf.cc (hunk 1 above)
Comment (cydrain) [reply src/index/ivf/ivf.cc:507]: updated

**Pull request merged**